File: Data Structures & Algorithms/rotting-fruit/submission-4.java
class Solution {
    private static final int[][] directions = new int[][]{{1,0},{-1,0},{0,1},{0,-1}};
    public int orangesRotting(int[][] grid) {
        Deque<int[]> q = new ArrayDeque<>();
        int ROWS = grid.length;
        int COLS = grid[0].length;
        int minutes=0;
        int fresh=0;
        for(int r=0;r<ROWS;r++){
            for(int c=0;c<COLS;c++){
                if(grid[r][c] == 2){
                    q.add(new int[]{r,c});
                }else if(grid[r][c] == 1){
                    fresh++;
                }
            }
        }

        while(!q.isEmpty() && fresh > 0){
            int len = q.size();
            minutes++;
            for(int i=0;i<len;i++){
                int[] cords = q.poll();
                int r = cords[0];
                int c = cords[1];
                for(int[] dir:directions){
                    int nr = r + dir[0];
                    int nc = c + dir[1];
                    if(nr < 0 || nr >= ROWS || nc <0 || nc >= COLS || grid[nr][nc]!=1){
                        continue;
                    }else{
                        q.add(new int[]{nr,nc});
                        grid[nr][nc] = 2;
                        fresh--;
                    }
                }
            }
        }
        return fresh == 0? minutes:-1;
    }
}

File: Data Structures & Algorithms/rotting-fruit/submission-4.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        time = 0
        ROWS = len(grid)
        COLS = len(grid[0])
        dirs = [[1,0],[-1,0],[0,1],[0,-1]]
        fresh = 0
        q = deque()
        rotten = 0

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 2:
                    rotten+=1
                    q.append((r,c))
                elif grid[r][c] == 1:
                    fresh+=1

        while q:
            if fresh == 0:
                break
            time+=1
            qLen = len(q)
            for _ in range(qLen):
                r,c = q.popleft()
                for dr,dc in dirs:
                    nr,nc = r+dr, c+dc
                    if nr < 0 or nr >= ROWS or nc<0 or nc>=COLS or grid[nr][nc] != 1:
                        continue 
                    grid[nr][nc] = 2
                    q.append((nr, nc))
                    fresh-=1
        return time if not fresh else -1 

        